docs/architecture-decision-records/2022-07-module-configuration.md

```diff
@@ -0,0 +1,95 @@
+willing to offer for the cache.
+
+This configuration is module-specific. Each module has different options.
+
+The Station communicates the configuration to the Module via environment
+variables passed to the module backend process.
+
+## Options Considered
+
+### (1) Each module provides config WebUI
+
+The initial idea was to let each Station Module implement its configuration
+screen as part of the module WebUI and persist the configuration in a file. The
+Station would provide the directory where to keep the config file.
+
+We quickly realised this was not optimal. The Saturn module is written in Go and
+expects users to provide all configuration options via environment variables in
+the spirit of [The Twelve-Factor App](https://12factor.net). Implementing a
+configuration layer with GUI requires a lot of work inside Saturn L2 that's not
+adding much value to Saturn:
+
+1. A REST API to read & write the configuration from the WebUI.
+2. Code to persist the configuration in a file. Also, handle possible conflicts
+   between the config file and environment variables.
+3. The config editor GUI.
+4. Support for changing configuration options at runtime with zero downtime.
+
+### (2) Module provides config schema; Station provides configuration WebUI
+
+A different approach is to mimic what VisualStudio Code offers to extensions:
+
+- Modules provide schema describing their configuration, and
+- Station builds a configuration GUI based on the module-provided schema.
+
+This creates a value proposition offered by the Station to future Module
+authors: modules don't have to implement a user-friendly configuration GUI; the
+Station will provide it for free.
+
+There are some caveats, though. Supporting complex configuration formats is a
+lot of work. VS Code has been evolving the configuration editor for years, and
+there are still edge cases where the user has to edit the configuration in the
+JSON file.
+
+Some edge cases to consider:
+
+- A config option is an array of primitive values (an array of strings, an array
+  of numbers)
+- A config option is an enum - there is a set of valid values to pick from.
+- Combine 1+2, and we have an array of enum values.
+- A config option is an object (a key-value map). In some cases, sub-options are
+  well defined - we know what the valid keys are and what value types each key
+  accepts. In other cases, this can be entirely dynamic.
+
+## Decision
+
+We have decided to implement the second option, where the Station provides the
+configuration GUI for Modules.
+
+We will start by supporting:
+
+- string keys mapped directly to environment variables
+- string or integer values
+- optional and required config options
+
+We will implement support for more complex configurations later.
+
+To keep our config schema easy to extend and forward-compatible, it should be a
+superset of
+[JSON Schema](https://json-schema.org/understanding-json-schema/reference/index.html).
+
```

Comment:
A discussion point: how will a Module provide the config schema?

I like the current module distribution model, where each module is a single binary file.

I propose requiring modules to provide a special CLI arg (e.g. `--config-schema`) for obtaining the configuration schema.

1. Saturn L2 can keep the config schema file alongside the Go source code.
2. Saturn L2 will use Go embed to bundle the config schema.
3. Saturn L2 detects `--config-schema` arg and prints config schema to stdout.
4. Station runs `saturn-l2 --config-schema` to get its config schema. We can do this as part of our build step. In the future, this can happen at module install time.

Thoughts?

Maybe this should go to a different ADR file?

Reply:
sgtm. How should the station/module handle breaking changes or updates? Say it wants to rename a config key, or add a config option, or update a default value?

We can diff the new config with the existing config and handle the various cases.

New key: Add to existing config
Missing key: Remove from existing config
Updated value: Update existing value except if the user set a value?

As for how to differentiate a renamed key vs a new key, I'm not sure. 
* Maybe it's easier to not allow it but we can't enforce it.
* Module author can "rename" by removing old key and adding a new key.
  *  But we need to retain the user specified value if it exists.

Reply:
> Say it wants to rename a config key, add a config option, or update a default value?

This is the classic API evolution and versioning problem, isn't it?

VS Code has the concept of deprecating config options. I guess the process goes like this:
1. Add a new option & deprecate the old one. Internally, use the deprecated option as a fallback if the new option is not provided.
2. Eventually, after some time passed, release a major version that removes the deprecated option.

I can imagine that the config schema can have a section describing migration paths from previous versions, akin to database migrations. Migration entries can be instructions like "delete this config field", "rename field A to B", and so on.

Anyhow, I think these complications are out of the scope of the initial release. I'll capture them in the ADR, though.

Reply:
I agree that this is outside the scope for now, but very good that you brought up these upgrade paths for the future.